src/lib/math/pcurves/pcurves.cpp

```diff
@@ -6,178 +6,177 @@
+   return nullptr;
 }
+#endif
 
-namespace sm2p256v1 {
-
-class Params final : public EllipticCurveParameters<
-  "FFFFFFFEFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFF00000000FFFFFFFFFFFFFFFF",
-  "FFFFFFFEFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFF00000000FFFFFFFFFFFFFFFC",
-  "28E9FA9E9D9F5E344D5A9E4BCF6509A7F39789F515AB8F92DDBCBD414D940E93",
-  "FFFFFFFEFFFFFFFFFFFFFFFFFFFFFFFF7203DF6B21C6052B53BBF40939D54123",
-  "32C4AE2C1F1981195F9904466A39C9948FE30BBFF2660BE1715A4589334C74C7",
-  "BC3736A2F4F6779C59BDCEE36B692153D0A9877CC62A474002DF32E52139F0A0"> {
-};
-
-class Curve final : public EllipticCurve<Params> {};
-
+#if !defined(BOTAN_HAS_PCURVES_FRP256V1)
+//static
+std::shared_ptr<const PrimeOrderCurve> PCurveInstance::frp256v1() {
+   return nullptr;
 }
+#endif
 
-// clang-format on
-
-}  // namespace
+#if !defined(BOTAN_HAS_PCURVES_SM2P256V1)
+//static
+std::shared_ptr<const PrimeOrderCurve> PCurveInstance::sm2p256v1() {
+   return nullptr;
+}
+#endif
 
 std::shared_ptr<const PrimeOrderCurve> PrimeOrderCurve::from_id(PrimeOrderCurveId id) {
    switch(id.code()) {
       case PrimeOrderCurveId::secp256r1:
-         return PrimeOrderCurveImpl<secp256r1::Curve>::instance();
+         return PCurveInstance::secp256r1();
       case PrimeOrderCurveId::secp384r1:
-         return PrimeOrderCurveImpl<secp384r1::Curve>::instance();
+         return PCurveInstance::secp384r1();
       case PrimeOrderCurveId::secp521r1:
-         return PrimeOrderCurveImpl<secp521r1::Curve>::instance();
+         return PCurveInstance::secp521r1();
       case PrimeOrderCurveId::secp256k1:
-         return PrimeOrderCurveImpl<secp256k1::Curve>::instance();
+         return PCurveInstance::secp256k1();
       case PrimeOrderCurveId::brainpool256r1:
-         return PrimeOrderCurveImpl<brainpool256r1::Curve>::instance();
+         return PCurveInstance::brainpool256r1();
       case PrimeOrderCurveId::brainpool384r1:
-         return PrimeOrderCurveImpl<brainpool384r1::Curve>::instance();
+         return PCurveInstance::brainpool384r1();
       case PrimeOrderCurveId::brainpool512r1:
-         return PrimeOrderCurveImpl<brainpool512r1::Curve>::instance();
+         return PCurveInstance::brainpool512r1();
       case PrimeOrderCurveId::frp256v1:
-         return PrimeOrderCurveImpl<frp256v1::Curve>::instance();
+         return PCurveInstance::frp256v1();
       case PrimeOrderCurveId::sm2p256v1:
-         return PrimeOrderCurveImpl<sm2p256v1::Curve>::instance();
+         return PCurveInstance::sm2p256v1();
    }
    return {};
 }
 
+std::vector<PrimeOrderCurveId> PrimeOrderCurveId::all() {
+   return {
+      PrimeOrderCurveId::secp256r1,
+      PrimeOrderCurveId::secp384r1,
+      PrimeOrderCurveId::secp521r1,
+      PrimeOrderCurveId::secp256k1,
+      PrimeOrderCurveId::brainpool256r1,
+      PrimeOrderCurveId::brainpool384r1,
+      PrimeOrderCurveId::brainpool512r1,
+      PrimeOrderCurveId::frp256v1,
+      PrimeOrderCurveId::sm2p256v1,
+   };
+}
```

Comment:
Perhaps this should only report the curves that are actually available in this build? I'm a bit on the fence about it, given that the curve-specific factories do exist regardless and just return a default value.

Reply:
Yeah, I'm not sure what's more helpful. The main usage of this is for the tests though; I don't really see why we would ever call this within the library. Maybe it's best to just move this list to the tests.

Also the tests still assume that all the curve are always available

```
            auto curve = Botan::PCurve::PrimeOrderCurve::from_id(id);

            if(!curve) {
               result.test_failure("ID exists but curve does not");
               continue;
            }
```

so I'll have to fix that regardless.

Reply:
Mhh, I think there's value in having this closer to the implementation, rather than moving it to the tests where it might get missed when adding curves. Perhaps just leave a docstring of warning?